server.js

```diff
@@ -7,20 +7,79 @@ const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
 mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
-//   PORT=9000 npm start
+const Thought = mongoose.model('Thought', {
+  message: {
+    type: String,
+    required: true,
+    minlength: 5,
+    maxlength: 140
+  },
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+})
 const port = process.env.PORT || 8080
 const app = express()
+const endPointList = require ('express-list-endpoints')
 
 // Add middlewares to enable cors and json body parsing
 app.use(cors())
 app.use(bodyParser.json())
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({ error: 'Service unavailable. Please try again later' })
+  }
+})
 
 // Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  if (!res) {
+    res
+    .status(404)
+    .send({ error: 'Oops! Something goes wrong. Try again later!' })
+  }
+  res.send(endPointList(app))
+})
+
+app.get('/thoughts', async (req, res) => {
+ try {
+   const thoughts = await Thought.find().sort({createdAt: 'desc'}).limit(20).exec()
+   res.json(thoughts)
+  } 
+  catch (err) {
+   res.status(404).json({ error: 'Page not found'})
+  }
+})
+
+app.post('/thoughts', async (req, res) => {
+  const {message} = req.body
+  try {
+    const newThought = await new Thought({message}).save()
+    res.status(201).json(newThought)
+  } catch (err) {
+    res.status(400).json({message: 'Could not save new thought', error: err.errors})
+  }
+})
+
+app.post('/thoughts/:thoughtId/like', async (req, res) => {
+  try {
+    const { thoughtId } = req.params
+    const newLike = await Thought.updateOne({ _id: thoughtId}, { $inc: { hearts: 1 } })
+    if (newLike) {
+      res.status(201).json(newLike)
+    } else {
+      res.status(404).json({ error: `thought message with id ${thoughtId} can not be found` })
+    }
+  } catch (err) {
+    res.status(400).json({ error: `Id ${thoughtId} is invalid` })
```

Comment:
I have tested your endpoint's in POSTMAN and I can't get the error handling to work for the like-endpoint

--> /thoughts/:thoughtId/like

Happy case works fine, so if I take a valid "thoughtId" I can POST a like.
--> https://whatmakesyouhappy.herokuapp.com/thoughts/5fffe618a382440023541e22/like

But if I send in an invalid "thoughtId" it's NOT working as expected.
--> https://whatmakesyouhappy.herokuapp.com/thoughts/5fffe618a382440023541e228888/like
--> https://whatmakesyouhappy.herokuapp.com/thoughts/8fffe6afa382440023541e26/like
--> https://whatmakesyouhappy.herokuapp.com/thoughts/7/like

For some reason it goes wrong in line 74 but I can't give an explanation why, perhaps you can bring this up with one of the code coaches why it doesn't work.

I also tried to declare a variable like you have done "newLike" but couldn't get the error handling to work so I decided to do it as VAN did like this.

```
// POST /thoughts/:thoughtId/like - Like a thought
app.post('/thoughts/:id/like', async (req, res) => {
  const { id } = req.params;

  try {
    // Success
    await Thought.updateOne({ _id: id }, { $inc: { hearts: 1 } });
    res.status(201).send();
  } catch (err) {
    res.status(400).json({
      message: `${COULD_NOT_SAVE_LIKE} ${id}`,
      error: err.errors
    });
  }
}); 
```